---
title: "venn-diagram"
author: "Valerie Gartner"
date: "11/20/2020"
output: html_document
---

(base) [vag15@hardac-login by-pop-vs-altogether-comparisons]$ cat vcf-compare_output.txt 
# This file was generated by vcf-compare.
# The command line was: vcf-compare(v0.1.14-12-gcdb80b8) by-pop-merged-data/by-pop-merged-chroms-snps-only.g.vcf.gz combined-joint-call-data/combined-joint-called-chroms-snps-only.g.vcf.gz
#
#VN 'Venn-Diagram Numbers'. Use `grep ^VN | cut -f 2-` to extract this part.
#VN The columns are: 
#VN        1  .. number of sites unique to this particular combination of files
#VN        2- .. combination of files and space-separated number, a fraction of sites in the file
VN	735	combined-joint-call-data/combined-joint-called-chroms-snps-only.g.vcf.gz (0.1%)
VN	2055	by-pop-merged-data/by-pop-merged-chroms-snps-only.g.vcf.gz (0.4%)
VN	502822	by-pop-merged-data/by-pop-merged-chroms-snps-only.g.vcf.gz (99.6%)	combined-joint-call-data/combined-joint-called-chroms-snps-only.g.vcf.gz (99.9%)
#SN Summary Numbers. Use `grep ^SN | cut -f 2-` to extract this part.
SN	Number of REF matches:	502822
SN	Number of ALT matches:	501592
SN	Number of REF mismatches:	0
SN	Number of ALT mismatches:	1230
SN	Number of samples in GT comparison:	0

```{r}
library(VennDiagram)
library(RColorBrewer)
library(ggplot2)
library(svglite)
```



```{r}
by_pop_cond <- 2055 + 502822
combined_cond <- 735 + 502822
overlap_area <- 502822

gvcf_plot <- draw.pairwise.venn(by_pop_cond, combined_cond, overlap_area, category = c("by_pop_cond", "combined_cond"), fill = c("blue", "red"),
	lty = "blank",
	cex = 2,
	cat.cex = 2,
	cat.pos = c(285, 105),
	cat.dist = 0.2,
	cat.just = list(c(-1, -1), c(1, 1)),
	ext.pos = 18,
	ext.dist = -0.15,
	ext.length = 0.55,
	ext.line.lwd = 2,
	ext.line.lty = "dashed",
	offset = 0.3)
grid.draw(gvcf_plot)
grid.newpage()

#saving as svg
#https://stackoverflow.com/questions/55097760/saving-a-venndiagram-output-to-svg
setwd("/Users/vgartner/Dropbox (Duke Bio_Ea)/vcf-joint-call/")
ggsave(gvcf_plot, file="my_plot.svg", device = "svg")

#tiff files don't open easily in Illustrator for editing
#tiff(filename = "Pairwise_Venn_diagram.tiff", compression = "lzw")
#grid.draw(gvcf_plot)
#dev.off()
```

#api singletons 
```{r}
bp_api <- 3 + 180
c_api <- 2 + 180
ov_api <- 180

sing_plot <- draw.pairwise.venn(bp_api, c_api, ov_api, fill = c("blue", "red"),
	lty = "blank",
	cex = 2,
	cat.cex = 2,
	cat.pos = c(285, 105),
	cat.dist = 0.2,
	cat.just = list(c(-1, -1), c(1, 1)),
	ext.pos = 18,
	ext.dist = -0.15,
	ext.length = 0.55,
	ext.line.lwd = 2,
	ext.line.lty = "dashed",
	offset = 0.3)

grid.draw(sing_plot)
grid.newpage()

setwd("/home/vgartner/wray-lab/joint-call/singletons/")
ggsave(sing_plot, file="api-singletons-venn-diagram.svg", device = "svg")


```

#By Chromosome Summary 
module load vcftools
module load tabix

$ for i in LT6356{12..27} ; do tabix -p vcf combined-joint-called-${i}.g.vcf.gz ; done

$ for i in LT6356{12..27} ; do echo vcf-compare combined-joint-call-data/combined-chroms/combined-joint-called-${i}.g.vcf.gz by-pop-merged-data/by-pop-chroms/by-pop-merged-${i}.g.vcf.gz ; done
```{r}

```